font/src/ft/mod.rs

```diff
@@ -242,14 +239,55 @@ impl FreeTypeRasterizer {
         pattern.add_style(style);
         pattern.add_pixelsize(self.pixel_size);
 
-        let font = fc::font_match(fc::Config::get_current(), &mut pattern)
+        self.query_font(pattern, desc)
+    }
+
+    fn query_font(&mut self, mut pattern: fc::Pattern, desc: &FontDesc) -> Result<FontKey, Error> {
+        let config = fc::Config::get_current();
+        let fonts = fc::font_sort(&config, &mut pattern, true)
             .ok_or_else(|| Error::MissingFont(desc.to_owned()))?;
-        self.face_from_pattern(&font).and_then(|pattern| {
+
+        let mut font_iter = fonts.into_iter();
+
+        let base_font = font_iter.next().ok_or_else(|| Error::MissingFont(desc.to_owned()))?;
+        let font_path = base_font.file(0).ok_or_else(|| Error::MissingFont(desc.to_owned()))?;
+
+        // Reload already loaded faces and drop their fallback faces
+        let font_key = if let Some(font_key) = self.keys.remove(&font_path) {
+            let fallback_list = self.fallback_lists.remove(&font_key).unwrap_or_default();
+
+            for font in fallback_list {
+                let font_path = font.file(0).unwrap();
+                if let Some(ff_key) = self.keys.get(&font_path) {
+                    // Our fallback font can be a primary font in some fallback list, so don't
+                    // remove it, since we will be reloading primary fonts all together
+                    if !self.fallback_lists.contains_key(&ff_key) {
+                        self.faces.remove(ff_key);
+                        self.keys.remove(&font_path);
+                    }
+                }
+            }
+
+            let _ = self.faces.remove(&font_key);
+            Some(font_key)
+        } else {
+            None
+        };
```

Comment:
You can't delete primary fonts, since your downstream client can request it without reload, etc, etc. Or you have a chance to skip cleaning and it'll break things. The logic is indeed a bit complex, but it was mostly due to our `FontKey` API and some internal structure.

Reply:
Yeah, makes sense. I'll probably try to play around with things a bit in a final review, I just wanted to leave a comment so everyone involved is aware that this might be problematic.

Reply:
I've tried to simplify it, but this logic seems like the most clear thing tbh.

- Delete key from keys set, but cache it for recreation
- Delete all fallbacks, however don't touch requested fonts
- Remove face for a primary font
- Return font key to recreate font properly later and don't break library users.